File: packages/ontology/src/action-preview.ts
import { canonicalizeJson, isCanonicalDigestHex, sha256Hex } from "./jcs.js";

export const ACTION_PREVIEW_SCHEMA = "zoen.action.preview.v1";
export const ACTION_PREVIEW_LOCALE = "pt-BR";

export type ActionPreviewValue =
  | { readonly kind: "bool"; readonly value: boolean }
  | { readonly kind: "decimal"; readonly value: string }
  | { readonly kind: "entity"; readonly value: string }
  | { readonly kind: "integer"; readonly value: string }
  | {
      readonly kind: "quantity";
      readonly amount: string;
      readonly unit: string;
    }
  | { readonly kind: "text"; readonly value: string };

export interface ActionPreviewInput {
  readonly id: string;
  readonly value: ActionPreviewValue;
}

export interface ActionPreviewDocument {
  readonly action: string;
  readonly canonicalPreviewText: string;
  readonly inputs: readonly ActionPreviewInput[];
  readonly locale: string;
  readonly resource: string;
  readonly schema: string;
}

export function buildActionPreviewDocument(input: {
  readonly actionId: string;
  readonly resourceId: string;
  readonly inputs: readonly ActionPreviewInput[];
}): ActionPreviewDocument {
  const inputs = [...input.inputs].sort((left, right) =>
    compareIds(left.id, right.id)
  );
  return {
    action: input.actionId,
    canonicalPreviewText: canonicalPreviewText(input.actionId, inputs),
    inputs,
    locale: ACTION_PREVIEW_LOCALE,
    resource: input.resourceId,
    schema: ACTION_PREVIEW_SCHEMA,
  };
}

export function canonicalPreviewText(
  actionId: string,
  inputs: readonly ActionPreviewInput[]
): string {
  switch (actionId) {
    case "personal.writeMemory":
      return `Vou guardar esta nota: ${textInput(inputs, "body") ?? ""}`;
    case "personal.createReminder":
      return `Vou criar este lembrete para ${textInput(inputs, "dueAt") ?? ""}: ${textInput(inputs, "body") ?? ""}`;
    default: {
      const label = actionLabel(actionId);
      const quantity = displayInput(inputs, "quantity");
      return quantity === undefined
        ? `Vou executar ${label}.`
        : `Vou executar ${label} com quantidade ${quantity}.`;
    }
  }
}

/**
 * SHA-256 of RFC 8785 JCS bytes. Matches zoen-engine::preview_hash.
 */
export function actionPreviewHash(document: ActionPreviewDocument): string {
  const canonical = canonicalizeJson(JSON.stringify(toWireDocument(document)));
  const digest = sha256Hex(canonical);
  if (!isCanonicalDigestHex(digest)) {
    throw new Error("preview hash is not lowercase SHA-256 hex");
  }
  return digest;
}

export function toWireDocument(document: ActionPreviewDocument): unknown {
  return {
    action: document.action,
    canonical_preview_text: document.canonicalPreviewText,
    inputs: document.inputs.map((input) => wireInput(input)),
    locale: document.locale,
    resource: document.resource,
    schema: document.schema,
  };
}

function wireInput(input: ActionPreviewInput): unknown {
  switch (input.value.kind) {
    case "bool":
      return { id: input.id, kind: "bool", value: input.value.value };
    case "decimal":
      return { id: input.id, kind: "decimal", value: input.value.value };
    case "entity":
      return { id: input.id, kind: "entity", value: input.value.value };
    case "integer":
      return { id: input.id, kind: "integer", value: input.value.value };
    case "quantity":
      return {
        amount: input.value.amount,
        id: input.id,
        kind: "quantity",
        unit: input.value.unit,
      };
    case "text":
      return { id: input.id, kind: "text", value: input.value.value };
    default: {
      const exhaustive: never = input.value;
      return exhaustive;
    }
  }
}

function compareIds(left: string, right: string): number {
  if (left < right) {
    return -1;
  }
  if (left > right) {
    return 1;
  }
  return 0;
}

function actionLabel(actionId: string): string {
  const parts = actionId.split(".");
  return parts.at(-1) ?? actionId;
}

function textInput(
  inputs: readonly ActionPreviewInput[],
  id: string
): string | undefined {
  const found = inputs.find((input) => input.id === id);
  return found?.value.kind === "text" ? found.value.value : undefined;
}

function displayInput(
  inputs: readonly ActionPreviewInput[],
  id: string
): string | undefined {
  const found = inputs.find((input) => input.id === id);
  if (found === undefined) {
    return undefined;
  }
  switch (found.value.kind) {
    case "bool":
      return String(found.value.value);
    case "decimal":
    case "entity":
    case "integer":
    case "text":
      return found.value.value;
    case "quantity":
      return `${found.value.amount} ${found.value.unit}`;
    default: {
      const exhaustive: never = found.value;
      return exhaustive;
    }
  }
}

File: crates/zoen-core/src/action_preview.rs
use std::fmt::Write as _;

use crate::{ActionId, ActionInput, ExactValue, ResourceId};

/// Schema id hashed into every Action preview document.
pub const ACTION_PREVIEW_SCHEMA: &str = "zoen.action.preview.v1";

/// Locale frozen for V1 preview text. A locale change needs a new schema.
pub const ACTION_PREVIEW_LOCALE: &str = "pt-BR";

/// One normalized Action input as it appears in the JCS preview document.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ActionPreviewInput {
    pub amount: Option<String>,
    pub id: String,
    pub kind: String,
    pub unit: Option<String>,
    pub value: Option<String>,
    pub value_bool: Option<bool>,
}

/// User-safe Action preview. Kernel hashes RFC 8785 JCS of this document.
///
/// Internal IDs (proposal, operation, claim, tenant, principal) stay out of
/// `canonical_preview_text`. `action` and `resource` bind the hash.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ActionPreviewDocument {
    pub action: String,
    pub canonical_preview_text: String,
    pub inputs: Vec<ActionPreviewInput>,
    pub locale: String,
    pub resource: String,
    pub schema: String,
}

impl ActionPreviewDocument {
    pub fn from_action(
        action_id: &ActionId,
        resource_id: &ResourceId,
        inputs: &[ActionInput],
    ) -> Self {
        let mut preview_inputs = inputs
            .iter()
            .map(ActionPreviewInput::from_action_input)
            .collect::<Vec<_>>();
        preview_inputs.sort_by(|left, right| left.id.cmp(&right.id));
        Self {
            action: action_id.as_str().to_owned(),
            canonical_preview_text: canonical_preview_text(action_id.as_str(), inputs),
            inputs: preview_inputs,
            locale: ACTION_PREVIEW_LOCALE.to_owned(),
            resource: resource_id.as_str().to_owned(),
            schema: ACTION_PREVIEW_SCHEMA.to_owned(),
        }
    }

    /// JSON object ready for RFC 8785 JCS. Key order is not significant.
    #[must_use]
    pub fn to_json(&self) -> String {
        let mut out = String::from("{");
        write_json_member(&mut out, "action", true);
        write_json_string(&mut out, &self.action);
        write_json_member(&mut out, "canonical_preview_text", false);
        write_json_string(&mut out, &self.canonical_preview_text);
        out.push_str(",\"inputs\":[");
        for (index, input) in self.inputs.iter().enumerate() {
            if index > 0 {
                out.push(',');
            }
            out.push_str(&input.to_json());
        }
        out.push(']');
        write_json_member(&mut out, "locale", false);
        write_json_string(&mut out, &self.locale);
        write_json_member(&mut out, "resource", false);
        write_json_string(&mut out, &self.resource);
        write_json_member(&mut out, "schema", false);
        write_json_string(&mut out, &self.schema);
        out.push('}');
        out
    }
}

impl ActionPreviewInput {
    fn from_action_input(input: &ActionInput) -> Self {
        match &input.value {
            ExactValue::Bool(value) => Self {
                amount: None,
                id: input.id.as_str().to_owned(),
                kind: "bool".to_owned(),
                unit: None,
                value: None,
                value_bool: Some(*value),
            },
            ExactValue::Decimal(value) => Self {
                amount: None,
                id: input.id.as_str().to_owned(),
                kind: "decimal".to_owned(),
                unit: None,
                value: Some(value.as_str().to_owned()),
                value_bool: None,
            },
            ExactValue::Entity(value) => Self {
                amount: None,
                id: input.id.as_str().to_owned(),
                kind: "entity".to_owned(),
                unit: None,
                value: Some(value.as_str().to_owned()),
                value_bool: None,
            },
            ExactValue::Integer(value) => Self {
                amount: None,
                id: input.id.as_str().to_owned(),
                kind: "integer".to_owned(),
                unit: None,
                value: Some(value.as_str().to_owned()),
                value_bool: None,
            },
            ExactValue::Quantity { amount, unit } => Self {
                amount: Some(amount.as_str().to_owned()),
                id: input.id.as_str().to_owned(),
                kind: "quantity".to_owned(),
                unit: Some(unit.as_str().to_owned()),
                value: None,
                value_bool: None,
            },
            ExactValue::Text(value) => Self {
                amount: None,
                id: input.id.as_str().to_owned(),
                kind: "text".to_owned(),
                unit: None,
                value: Some(value.clone()),
                value_bool: None,
            },
        }
    }

    fn to_json(&self) -> String {
        let mut out = String::from("{");
        if let Some(amount) = &self.amount {
            write_json_member(&mut out, "amount", true);
            write_json_string(&mut out, amount);
        }
        write_json_member(&mut out, "id", self.amount.is_none());
        write_json_string(&mut out, &self.id);
        write_json_member(&mut out, "kind", false);
        write_json_string(&mut out, &self.kind);
        if let Some(unit) = &self.unit {
            write_json_member(&mut out, "unit", false);
            write_json_string(&mut out, unit);
        }
        if let Some(value) = &self.value {
            write_json_member(&mut out, "value", false);
            write_json_string(&mut out, value);
        }
        if let Some(value) = self.value_bool {
            write_json_member(&mut out, "value", false);
            out.push_str(if value { "true" } else { "false" });
        }
        out.push('}');
        out
    }
}

/// Deterministic PT-BR preview. Does not mention proposal, operation, or claim IDs.
#[must_use]
pub fn canonical_preview_text(action_id: &str, inputs: &[ActionInput]) -> String {
    match action_id {
        "personal.writeMemory" => {
            format!(
                "Vou guardar esta nota: {}",
                text_input(inputs, "body").unwrap_or("")
            )
        }
        "personal.createReminder" => {
            format!(
                "Vou criar este lembrete para {}: {}",
                text_input(inputs, "dueAt").unwrap_or(""),
                text_input(inputs, "body").unwrap_or("")
            )
        }
        other => {
            let label = action_label(other);
            if let Some(quantity) = display_input(inputs, "quantity") {
                format!("Vou executar {label} com quantidade {quantity}.")
            } else {
                format!("Vou executar {label}.")
            }
        }
    }
}

fn action_label(action_id: &str) -> &str {
    action_id.rsplit('.').next().unwrap_or(action_id)
}

fn text_input<'a>(inputs: &'a [ActionInput], id: &str) -> Option<&'a str> {
    inputs.iter().find_map(|input| {
        if input.id.as_str() != id {
            return None;
        }
        match &input.value {
            ExactValue::Text(value) => Some(value.as_str()),
            _ => None,
        }
    })
}

fn display_input(inputs: &[ActionInput], id: &str) -> Option<String> {
    inputs.iter().find_map(|input| {
        if input.id.as_str() != id {
            return None;
        }
        Some(display_exact_value(&input.value))
    })
}

fn display_exact_value(value: &ExactValue) -> String {
    match value {
        ExactValue::Bool(value) => value.to_string(),
        ExactValue::Decimal(value) => value.as_str().to_owned(),
        ExactValue::Entity(value) => value.as_str().to_owned(),
        ExactValue::Integer(value) => value.as_str().to_owned(),
        ExactValue::Quantity { amount, unit } => {
            format!("{} {}", amount.as_str(), unit.as_str())
        }
        ExactValue::Text(value) => value.clone(),
    }
}

fn write_json_member(out: &mut String, key: &str, first: bool) {
    if !first {
        out.push(',');
    }
    write_json_string(out, key);
    out.push(':');
}

fn write_json_string(out: &mut String, value: &str) {
    out.push('"');
    for character in value.chars() {
        match character {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{0008}' => out.push_str("\\b"),
            '\u{000c}' => out.push_str("\\f"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            character if u32::from(character) < 0x20 => {
                let _ = write!(out, "\\u{:04x}", u32::from(character));
            }
            character => out.push(character),
        }
    }
    out.push('"');
}
